base/strings/annotated.jl

```diff
@@ -386,3 +386,62 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
 Get all annotations of `chr`.
 """
 annotations(c::AnnotatedChar) = c.annotations
+
+## AnnotatedIOBuffer
+
+struct AnnotatedIOBuffer <: IO
+    io::IOBuffer
+    annotations::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}
+    dict::ImmutableDict{Symbol, Any}
+end
+
+AnnotatedIOBuffer(io::IOBuffer, dict::ImmutableDict = ImmutableDict{Symbol,Any}()) =
+    AnnotatedIOBuffer(io, Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}(), dict)
+
+function AnnotatedIOBuffer(io::IOBuffer, (k, v)::Pair{Symbol, Any}, kvs::Pair{Symbol, Any}...)
+    dict = ImmutableDict{Symbol,Any}(k, v)
+    for (k, v) in kvs
+        dict = ImmutableDict{Symbol,Any}(dict, k, v)
+    end
+    AnnotatedIOBuffer(io, dict)
+end
+
+AnnotatedIOBuffer() = AnnotatedIOBuffer(IOBuffer())
+
+function show(io::IO, annio::AnnotatedIOBuffer)
+    show(io, AnnotatedIOBuffer)
+    print(io, '(', annio.io.size, " bytes)")
+end
+
+position(io::AnnotatedIOBuffer) = position(io.io)
+lock(io::AnnotatedIOBuffer) = lock(io.io)
+unlock(io::AnnotatedIOBuffer) = unlock(io.io)
+get(io::AnnotatedIOBuffer, key, default) = key == :color || get(io.dict, key, default)
+displaysize(io::AnnotatedIOBuffer) =
+    if haskey(io.dict, :displaysize) io.dict[:displaysize]::Tuple{Int,Int} else displaysize(io.io) end
+
+function write(io::AnnotatedIOBuffer, astr::Union{AnnotatedString, SubString{<:AnnotatedString}})
+    astr = AnnotatedString(astr)
+    offset = position(io.io)
+    for (region, annot) in astr.annotations
+        start, stop = first(region), last(region)
+        push!(io.annotations, (start+offset:stop+offset, annot))
+    end
+    write(io.io, astr)
+end
+write(io::AnnotatedIOBuffer, achr::AnnotatedChar) = write(aio, AnnotatedString(achr))
+write(io::AnnotatedIOBuffer, x::AbstractString) = write(io.io, x)
+write(io::AnnotatedIOBuffer, x::UInt8) = write(io.io, x)
+write(io::AnnotatedIOBuffer, x::AbstractString) = write(io.io, x)
+write(io::AnnotatedIOBuffer, s::Union{SubString{String}, String}) = write(io.io, x)
+
+function take!(aiob::AnnotatedIOBuffer)
+    str = String(take!(aiob.io))
+    annot = copy(aiob.annotations)
+    empty!(aiob.annotations)
+    seekstart(aiob.io)
+    str, annot
+end
+
+AnnotatedString((str, annots)::Tuple{<:AbstractString, Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}}) =
+    AnnotatedString(str, annots)
```

Comment:
This seems like a strange API decision; why not return an `AnnotatedString`?

Reply:
This enables `AnnotatedString(take!(::AnnotatedIOBuffer))`. I considered having `AnnotatedString(::AnnotatedIOBuffer)`, but I don't like that it mutates the state of the `AnnotatedIOBuffer` within the constructor, that feels icky to me.

Reply:
It occurs to me that maybe `read(::AnnotatedIOBuffer, AnnotatedString)` may also be reasonable.

Reply:
One worry I have about the current approach is that it exposes the annotations field which is otherwise internal.

I like `read(::AnnotatedIOBuffer, AnnotatedString)`

`take!(::IOBuffer)` returns a vector, not an AbstractString. It seems reasonable to be consistent with that behavior, but also idk how to return a vector analogue that contains annotation information. I don't love `AnnotatedVector`—that feels like a bit much.

Reply:
Changed :)